File: examples/Clojure/lpserver.clj
(ns lpserver
  (:require [zeromq.zmq :as zmq]))

(defn -main []
  (let [ctx (zmq/zcontext)
        server (doto (zmq/socket ctx :rep)
                 (zmq/bind "tcp://*:5555"))]
    (try
      (doseq [cycles (iterate inc 1)]
        (let [req (zmq/receive-str server)]
          (cond
           (and (> cycles 3) (zero? (rand-int 3)))
           (do (println "I: simulating a crash")
               (throw (Exception. "crashing")))
           (and (> cycles 3) (zero? (rand-int 3)))
           (do (println "I: simulating CPU overload")
               (Thread/sleep 2000))
           :else nil)
          (println (format "I: normal request (%s)" req))
          (Thread/sleep 1000)
          (zmq/send-str server req)))
      (catch Exception _ nil))
    (zmq/destroy ctx)))

File: examples/CSharp/lpserver.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;

using ZeroMQ;

namespace Examples
{
	static partial class Program
	{
		//
		// Lazy Pirate server
		// Binds REP socket to tcp://*:5555
		// Like hwserver except:
		// - echoes request as-is
		// - randomly runs slowly, or exits to simulate a crash.
		//
		// Author: metadings
		//

		public static void LPServer(string[] args)
		{
			using (var context = new ZContext())
			using (var responder = new ZSocket(context, ZSocketType.REP))
			{
				responder.Bind("tcp://*:5555");

				ZError error;
				int cycles = 0;
				var rnd = new Random();

				while (true)
				{
					ZMessage incoming;
					if (null == (incoming = responder.ReceiveMessage(out error)))
					{
						if (error == ZError.ETERM)
							return;	// Interrupted
						throw new ZException(error);
					}
					using (incoming)
					{
						++cycles;

						// Simulate various problems, after a few cycles
						if (cycles > 16 && rnd.Next(16) == 0)
						{
							Console.WriteLine("I: simulating a crash");
							break;
						}
						else if (cycles > 4 && rnd.Next(4) == 0)
						{
							Console.WriteLine("I: simulating CPU overload");
							Thread.Sleep(1000);
						}

						Console.WriteLine("I: normal request ({0})", incoming[0].ReadInt32());

						Thread.Sleep(1); // Do some heavy work

						responder.Send(incoming);
					}
				}
			}
		}
	}
}